src/main/java/org/folio/inventory/dataimport/handlers/actions/UpdateHoldingEventHandler.java

```diff
@@ -201,9 +207,28 @@ private void prepareDataAndReInvokeCurrentHandler(DataImportEventPayload dataImp
     });
   }
 
-  private void constructDataImportEventPayload(CompletableFuture<DataImportEventPayload> future, DataImportEventPayload dataImportEventPayload, HoldingsRecord holding) {
-    dataImportEventPayload.getContext().put(HOLDINGS.value(), Json.encodePrettily(holding));
-    future.complete(dataImportEventPayload);
-    future.complete(dataImportEventPayload);
+  private void constructDataImportEventPayload(CompletableFuture<DataImportEventPayload> future, DataImportEventPayload dataImportEventPayload, ItemCollection itemCollection, HoldingsRecord holding) {
+    HashMap<String, String> payloadContext = dataImportEventPayload.getContext();
+    payloadContext.put(HOLDINGS.value(), Json.encodePrettily(holding));
+    if(payloadContext.containsKey(ITEM.value())) {
+      updateDataImportEventPayloadItem(future, dataImportEventPayload, itemCollection);
+    } else {
+      future.complete(dataImportEventPayload);
+    }
+  }
+
+  private void updateDataImportEventPayloadItem(CompletableFuture<DataImportEventPayload> future, DataImportEventPayload dataImportEventPayload, ItemCollection itemCollection) {
+    JsonObject oldItemAsJson = new JsonObject(dataImportEventPayload.getContext().get(ITEM.value()));
+    String itemId = oldItemAsJson.getString(ITEM_ID_HEADER);
+    itemCollection.findById(itemId, findResult -> {
+      if(Objects.nonNull(findResult)) {
+        JsonObject itemAsJson = new JsonObject(ItemUtil.mapToMappingResultRepresentation(findResult.getResult()));
+        dataImportEventPayload.getContext().put(ITEM.value(), Json.encode(itemAsJson));
+      }
+      future.complete(dataImportEventPayload);
+    }, failure -> {
+      LOGGER.error("Cannot get actual Item after successfully updating holdings, by ITEM id: {} - {}, status code {}", itemId, failure.getReason(), failure.getStatusCode());
+      future.complete(dataImportEventPayload);
```

Comment:
Can you add a comment why this successfully completes the future? Or should this be `completeExceptionally`?

Reply:
Hi @julianladisch . This may be affected holding update proces in cases when we have in payload not yet created item. I'm not sure about it and better would be waiting feedback from other Folijet folks. Thank you.

Reply:
Thank you, @julianladisch! Fixed.